Fixes "Mutations and traits are a global list"

BevapDin opened the pull request:
Fixes #11143
- static mutation data (describing the mutation, not related to a specific creature in the game) is in `mutation_branch`.
- mutation status as part of a creature (whether it's activated by the creature, the invlet, ...) is in `trait_data`.
- The `mutation_data` map is hidden, but a const-qualified version of it can be accessed through some static functions in the class. This ensures that global mutation data is not modified after loading it (as it would affect all creatures).
- `Character::my_mutations` is now a map of `trait_data` objects, those objects contain the trait invlet from `Character::trait_keys` and the per-creature data of mutations (e.g. whether it's on).
- I added consistency checking of mutation data and consequently fixed some wrong mutation ids.

Questions:
- Rename `mutation_branch` to `mutation_type`?

KA101 commented on src/character.h:
```diff
@@ -290,9 +291,30 @@ class Character : public Creature
+            /** Key to select the mutation in the UI. */
+            char key = ' ';
+            /** TODO: I don't know who I am - comment me. */
+            int charge = 0;
```
I'm gonna take a wild guess and say that charge and cooldown are similar.

KA101 commented on src/mutation.cpp:
```diff
@@ -40,40 +44,19 @@ void Character::toggle_trait(const std::string &flag)
+        my_mutations.erase( iter );
     }
+    // TODO: why does this not call mutation_loss_effect / mutation_effect?
+    recalc_sight_limits();
```
I'm thinking it's because those are handled in gaining or removing a mutation, last I knew.  Currently the only stuff that gets specifically toggled rather than mutate_towards'd is thresholds and profession traits; mutate_towards calls the effects.

So yeah, probably should get 'em for future-proofing.

KA101 commented on src/mutation.cpp:
```diff
@@ -308,66 +291,65 @@ void Character::mutation_loss_effect(std::string mut)
+    if (tdata.powered && tdata.charge > 0) {
         // Already-on units just lose a bit of charge
-        traits[mut].charge--;
+        tdata.charge--;
```
Yeah, here's charge.  It's supposed to handle timing the active traits that take X resources per T time.

BevapDin commented:
I added some documentation and removed the TODOs.

About mutation_loss_effect / mutation_effect in `toggle_mutation`: that is probably fine. It is called for example from `mutation_towards` and the code there calls `mutation_loss_effect` / `mutation_effect`, so it's fine there.

It might be useful to implement simple `add_mutation` / `remove_mutation` functions to be used in iuse.cpp (the marloss / threshold stuff) and in mission_start.cpp. Actually if you get the same type of mission (that gives the PROF_FED trait) twice, you'll lose the PROF_FED traits again - because it *toggles* the mutation (-:

KA101 commented:
Yeah, the quests need some attention and I'll grant that add/remove would be handy.

I'm kinda waiting on this one, both thanks to large/0.C freeze and having finally worked up a fix for that multi-cancellation issue (though I'm thinking a while/do might be more effective).

kevingranade commented on src/character.h:
```diff
@@ -290,9 +291,34 @@ class Character : public Creature
-        
+        struct trait_data : public JsonSerializer, public JsonDeserializer {
+            /** Key to select the mutation in the UI. */
+            char key = ' ';
```
This belongs in the UI code, stashing it in the referent like this has caused a lot of pain over in the inventory code.

BevapDin replied:
So you want a separate map for the char key to mutation mapping? Should that be done for bionics as well?

It seems a bit pointless and error prone, one has to keep both maps synchronized all the time.

KA101 replied:
You two figure that out and let me know, please.  ;-)
